src/transaction/log_replication_atomic.cpp

```diff
@@ -402,41 +407,114 @@ namespace cublog
 	assert_release (false);
       }
 
-#if !defined (NDEBUG)
-    /* there will be no duplicated object in the map */
-    auto [begin, end] = m_locked_classes.equal_range (trid);
-    if (std::count_if (begin, end, [classoid] (const auto &it)
-    {
-      return it.second == *classoid;
-    }) > 0)
-    {
-      assert (false);
-    }
-#endif
+    m_locked_classes[trid].emplace (*classoid);
+  }
+
+  void
+  atomic_replicator::bookkeep_classname (cubthread::entry &thread_entry, const OID *classoid)
+  {
+    char *classname = nullptr;
 
-    m_locked_classes.emplace (trid, *classoid);
+    assert (!OID_ISNULL (classoid) && !OID_ISTEMP (classoid));
+
+    (void) heap_get_class_name (&thread_entry, classoid, &classname);
+    if (classname != nullptr)
+      {
+	m_classname_map.emplace (*classoid, std::move (std::string (classname)));
+      }
+    else
+      {
+	/* if classname == NULL, then it is CREATE TABLE/VIEW case */
+      }
   }
 
   void
-  atomic_replicator::discard_caches_for_ddl (cubthread::entry &thread_entry, const TRANID trid)
+  atomic_replicator::update_classname_cache_for_ddl (cubthread::entry &thread_entry, const OID *classoid)
   {
-    assert (m_locked_classes.count (trid) > 0);
+    /* update locator_Mht_classnames only if needed */
+    char *classname = nullptr;
+    int error_code = NO_ERROR;
 
-    /* TODO:
-     * This is to update locator_Mht_classnames which contains class names,
-     * and it will be replaced with a function to update specific class name only if needed.
-     */
-    locator_initialize (&thread_entry);
+    auto it = m_classname_map.find (*classoid);
+
+    (void) heap_get_class_name (&thread_entry, classoid, &classname);
+    if (classname != nullptr)
+      {
+	/* Memory for classname is allocated in heap_get_class_name ().
+	 * The allocated memory will be used within an entry of locator_Mht_classnames.
+	 * (in locator_initialize () called from locator_put_classname_entry () and locator_update_classname_entry ())
+	 * And when that entry is released, this memory will be freed as well
+	 */
+	if (it == m_classname_map.end ())
+	  {
+	    error_code = locator_put_classname_entry (&thread_entry, classname, classoid);
+	    if (error_code != NO_ERROR)
+	      {
+		_er_log_debug (ARG_FILE_LINE, "[REPL_DDL] Failed to put entry for (%s) to locator_Mht_classnames for classoid\n",
+			       classname);
+	      }
+	  }
+	else
+	  {
+	    if (it->second != classname)
+	      {
+		error_code = locator_update_classname_entry (&thread_entry, it->second.c_str(), classname);
+		if (error_code != NO_ERROR)
+		  {
+		    _er_log_debug (ARG_FILE_LINE, "[REPL_DDL] Failed to update entry for (%s) to (%s) in locator_Mht_classnames\n",
+				   it->second.c_str (), classname);
+		  }
+	      }
+
+	  }
+      }
+    else
+      {
+	/* if classname == nullptr, then it is DROP TABLE/VIEW case */
```

Comment:
```suggestion
	/* if classname == nullptr, the class (table/view) is dropped during the transaction. */
```
The transaction include not only DROP, but other DDLs.

Reply:
Oh, yes. Thanks.